pkg/providers/vsphere/validator.go

```diff
@@ -385,3 +403,184 @@ func (v *Validator) validateControlPlaneIpUniqueness(spec *Spec) error {
 	}
 	return nil
 }
+
+func (v *Validator) validateUserPrivs(ctx context.Context, spec *Spec, vuc *config.VSphereUserConfig) error {
+	controlPlaneMachineConfig := spec.controlPlaneMachineConfig()
+
+	privObjs := []map[string]string{
+		// validate global root priv settings are correct
+		{
+			"type":         vsphereTypeFolder,
+			"privsContent": config.VSphereGlobalPrivsFile,
+			"path":         vsphereRootPath,
+		},
+
+		// validate object-level priv settings are correct
+		{
+			"type":         vsphereTypeDatastore,
+			"privsContent": config.VSphereUserPrivsFile,
+			"path":         controlPlaneMachineConfig.Spec.Datastore,
+		},
+		{
+			"type":         vsphereTypeResourcePool,
+			"privsContent": config.VSphereUserPrivsFile,
+			"path":         controlPlaneMachineConfig.Spec.ResourcePool,
+		},
+		{
+			"type":         vsphereTypeNetwork,
+			"privsContent": config.VSphereUserPrivsFile,
+			"path":         spec.datacenterConfig.Spec.Network,
+		},
+
+		// validate Administrator role (all privs) on VM folder and Template folder
+		{
+			"type":         vsphereTypeFolder,
+			"privsContent": config.VSphereAdminPrivsFile,
+			"path":         controlPlaneMachineConfig.Spec.Folder,
+		},
+		// Not implementing a check on the template because I'm not sure if we always require/provide/allow template permissions
+		// https://anywhere.eks.amazonaws.com/docs/reference/vsphere/vsphere-preparation/#deploy-an-ova-template
+		// {
+		// 	"type": "VirtualMachine",
+		// 	"filepath": config.VSphereAdminPrivsFilepath,
+		// 	"path": controlPlaneMachineConfig.Spec.Template,
+		// },
+	}
+
+	host := spec.datacenterConfig.Spec.Server
+	datacenter := spec.datacenterConfig.Spec.Datacenter
+
+	vsc, err := v.newVSphereClient(
+		ctx,
+		host,
+		vuc.EksaVsphereCPUsername,
+		vuc.EksaVsphereCPPassword,
+		spec.datacenterConfig.Spec.Insecure,
+		datacenter,
+	)
+	if err != nil {
+		return err
+	}
+
+	return v.validatePrivs(ctx, privObjs, vsc)
+}
+
+func (v *Validator) validateCSIUserPrivs(ctx context.Context, spec *Spec, vuc *config.VSphereUserConfig) error {
+	controlPlaneMachineConfig := spec.controlPlaneMachineConfig()
+
+	privObjs := []map[string]string{
+		{ // CNS-Datastore role
+			"type":         vsphereTypeDatastore,
+			"privsContent": config.VSphereCnsDatastorePrivsFile,
+			"path":         controlPlaneMachineConfig.Spec.Datastore,
+		},
+		{ // CNS-SEARCH-AND-SPBM role
+			"type":         vsphereTypeFolder,
+			"privsContent": config.VSphereCnsSearchAndSpbmPrivsFile,
+			"path":         vsphereRootPath,
+		},
+		{ // CNS-VM role
+			"type":         vsphereTypeFolder,
+			"privsContent": config.VSphereCnsVmPrivsFile,
+			"path":         controlPlaneMachineConfig.Spec.Folder,
+		},
+		// CNS-HOST-CONFIG-STORAGE I don't know what this applies to yet
```

Comment:
Ask Vignesh or Abhinav about where this should be applied.

Reply:
Based on this conversation, I think this role only needs to be applied to the datastore.